tsdb/meta.go

```diff
@@ -846,69 +827,140 @@ func (m *Measurement) TagSets(dimensions []string, condition influxql.Expr) ([]*
-	// |          |----------+----------+-----------------------+-----------------------+
-	// |          | <nil>    | <nil>    | false OR false        | false                 |
-	// |          |----------+----------+-----------------------+-----------------------+
-	// |          | <l-expr> | <r-expr> | <l-expr> OR <r-expr>  | <l-expr> OR <r-expr>  |
-	// +----------+----------+----------+-----------------------+-----------------------+
-	// |          | <nil>    | <r-expr> | false AND <r-expr>    | false*                |
-	// |          |----------+----------+-----------------------+-----------------------+
-	// | AND      | <l-expr> | <nil>    | <l-expr> AND false    | false                 |
-	// |          |----------+----------+-----------------------+-----------------------+
-	// |          | <nil>    | <nil>    | false AND false       | false                 |
-	// |          |----------+----------+-----------------------+-----------------------+
-	// |          | <l-expr> | <r-expr> | <l-expr> AND <r-expr> | <l-expr> AND <r-expr> |
-	// +----------+----------+----------+-----------------------+-----------------------+
-	// *literal false filters and series IDs should be excluded from the results
+// intersectSeriesFilters performs an intersection for two sets of ids and filter expressions.
+func intersectSeriesFilters(lids, rids SeriesIDs, lfilters, rfilters FilterExprs) (SeriesIDs, FilterExprs) {
+	// We only want to allocate a slice and map of the smaller size.
+	var ids []uint64
+	if len(lids) > len(rids) {
+		ids = make([]uint64, 0, len(rids))
+	} else {
+		ids = make([]uint64, 0, len(lids))
+	}
 
-	for _, id := range ids {
-		// Get LHS and RHS filter expressions for this series ID.
-		lfilter, rfilter := lfilters[id], rfilters[id]
+	var filters FilterExprs
+	if len(lfilters) > len(rfilters) {
+		filters = make(FilterExprs, len(rfilters))
+	} else {
+		filters = make(FilterExprs, len(lfilters))
+	}
 
-		// Set filter to false if either LHS or RHS expressions were nil.
-		if lfilter == nil {
-			lfilter = &influxql.BooleanLiteral{Val: false}
-		}
-		if rfilter == nil {
-			rfilter = &influxql.BooleanLiteral{Val: false}
+	// They're in sorted order so advance the counter as needed.
+	// This is, don't run comparisons against lower values that we've already passed.
+	var i, j int
+	for i < len(lids) && j < len(rids) {
+		if lids[i] == rids[j] {
+			ids = append(ids, lids[i])
+
+			var expr influxql.Expr
+			lfilter := lfilters[lids[i]]
+			rfilter := rfilters[rids[j]]
+
+			if lfilter != nil && rfilter != nil {
+				be := &influxql.BinaryExpr{
+					Op:  influxql.AND,
+					LHS: lfilter,
+					RHS: rfilter,
+				}
+				expr = influxql.Reduce(be, nil)
+			} else if lfilter != nil {
+				expr = lfilter
+			} else if rfilter != nil {
+				expr = rfilter
+			}
+
+			if expr != nil {
+				filters[lids[i]] = expr
+			}
+
+			i++
+			j++
+		} else if lids[i] < rids[j] {
+			i++
+		} else {
+			j++
 		}
+	}
+	return ids, filters
+}
```

Comment:
You can simplify this and remove `i`/`j` by using the first element of `lids` and `rids` and reslicing.

e.g.

```go
for len(lids) > 0 && len(rids) > 0 {
	lid, rid := lids[0], rids[0]
	if lid == rid {
		// ...compare

		lids, rids = lids[1:], rids[1:]
	} else if lid < rid {
		lids = lids[1:]
	} else {
		rids = rids[1:]
	}
}
```